stdlib/3/unittest/mock.pyi

```diff
@@ -221,59 +249,113 @@ class _patch_dict:
 class _patcher:
     TEST_PREFIX: str
     dict: Type[_patch_dict]
-    @overload
-    def __call__(  # type: ignore
-        self,
-        target: Any,
-        *,
-        spec: Optional[Any] = ...,
-        create: bool = ...,
-        spec_set: Optional[Any] = ...,
-        autospec: Optional[Any] = ...,
-        new_callable: Optional[Any] = ...,
-        **kwargs: Any,
-    ) -> _patch[MagicMock]: ...
-    # This overload also covers the case, where new==DEFAULT. In this case, the return type is _patch[Any].
-    # Ideally we'd be able to add an overload for it so that the return type is _patch[MagicMock],
-    # but that's impossible with the current type system.
-    @overload
-    def __call__(
-        self,
-        target: Any,
-        new: _T,
-        spec: Optional[Any] = ...,
-        create: bool = ...,
-        spec_set: Optional[Any] = ...,
-        autospec: Optional[Any] = ...,
-        new_callable: Optional[Any] = ...,
-        **kwargs: Any,
-    ) -> _patch[_T]: ...
-    @overload
-    def object(  # type: ignore
-        self,
-        target: Any,
-        attribute: str,
-        *,
-        spec: Optional[Any] = ...,
-        create: bool = ...,
-        spec_set: Optional[Any] = ...,
-        autospec: Optional[Any] = ...,
-        new_callable: Optional[Any] = ...,
-        **kwargs: Any,
-    ) -> _patch[MagicMock]: ...
-    @overload
-    def object(
-        self,
-        target: Any,
-        attribute: str,
-        new: _T = ...,
-        spec: Optional[Any] = ...,
-        create: bool = ...,
-        spec_set: Optional[Any] = ...,
-        autospec: Optional[Any] = ...,
-        new_callable: Optional[Any] = ...,
-        **kwargs: Any,
-    ) -> _patch[_T]: ...
+    if sys.version_info >= (3, 8):
+        @overload
+        def __call__(  # type: ignore
+            self,
+            target: Any,
+            *,
+            spec: Optional[Any] = ...,
+            create: bool = ...,
+            spec_set: Optional[Any] = ...,
+            autospec: Optional[Any] = ...,
+            new_callable: Optional[Any] = ...,
+            **kwargs: Any,
+        ) -> _patch[Union[MagicMock, AsyncMock]]: ...
```

Comment:
Also, returning unions is always problematic, since it can force callers to use `isinstance` checks. One workaround is to use `_patch[Any]`, but maybe we can come up with a better solution using overloads?

Reply:
I though about this too. Unfortunately there doesn't seem to be a way to distinguish between patching a normal function -> `MagicMock` and a coroutine -> `AsyncMock`. At least I couldn't come up with any. Overloading doesn't work, since the args are the same.

I'm not sure returning returning `Any` is necessary. Since both inherit from `NonCallableMock` which has
`def __getattr__(self, name: str) -> Any: ...`, calling a function or variable that doesn't exist wouldn't raise an error with type checkers. It's true that it then can't be validated, but that would also be the case when using `Any`. With union we at least have some information for IDEs to provide suggestions.

Reply:
Makes sense.